### PR TITLE
Refactor proof rule documentation

### PR DESCRIPTION
**Pull request by nafur**
This PR starts to refactor the proof rule comments so that they generate proper sphinx documentation.
It sets up doxygen to include the `proof_rule.h`, includes some useful configuration for mathjax and provides proper documentation for all core rules and Boolean rules.

---

The documentation is at https://cvc5.github.io/docs-ci/docs-pr8303/proofs/proof_rules.html

### CODE REVIEW

File: src/proof/proof_rule.h [hunk 1]
```diff
@@ -37,595 +50,955 @@ namespace cvc5 {
...
-  //   (SCOPE (ASSUME F) :args F)
-  // has the conclusion (=> F F) and has no free assumptions. More generally, a
-  // proof with no free assumptions always concludes a valid formula.
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Scope (a binder for assumptions)**
+   *
+   * .. math::
+   *
+   *   \inferruleSC{F\neq\bot \mid F_1 \dots F_n}{(F_1 \land \dots \land F_n)
```
Comment (ajreynol) [inline src/proof/proof_rule.h:88]:
```suggestion
   *   \inferruleSC{F \mid F_1 \dots F_n}{(F_1 \land \dots \land F_n)
```

File: src/proof/proof_rule.h [hunk 2]
```diff
@@ -37,595 +50,955 @@ namespace cvc5 {
...
+   *   \Rightarrow F}{if $F\neq\bot$} \textrm{ or } \inferruleSC{F=\bot \mid F_1
```
Comment (ajreynol) [inline src/proof/proof_rule.h:89]:
```suggestion
   *   \Rightarrow F}{if $F\neq\bot$} \textrm{ or } \inferruleSC{F \mid F_1
```

**Review by ajreynol — CHANGES_REQUESTED** [review]
Initial minor comment.

File: src/proof/proof_rule.h [hunk 3]
```diff
@@ -23,9 +23,22 @@
 namespace cvc5 {
 
 /**
+ * \verbatim embed:rst:leading-asterisk
```
Comment (ajreynol) [inline src/proof/proof_rule.h:26]: Can you write the link in text of the html someone should look at to see the nice version of this header?

Comment (nafur) [reply src/proof/proof_rule.h:26]: It's in the initial comment of this PR: https://cvc5.github.io/docs-ci/docs-pr8303/proofs/proof_rules.html

Comment (ajreynol) [reply src/proof/proof_rule.h:26]: I'm saying that in the future, if someone looks at this header, we need a link to where they should look.

In particular, I'm concerned that after this PR the link from our IJCAR submission will point people to this document, and they will be unaware they need to go the html.

**Review by HanielB — CHANGES_REQUESTED** [review]

File: src/proof/proof_rule.h [hunk 4]
```diff
@@ -37,606 +53,956 @@ namespace cvc5 {
...
-  // where F' and G' are the result of each side of the equation above. Here,
-  // original forms are used in a similar manner to MACRO_SR_PRED_INTRO above.
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Builtin theory -- Substitution + Rewriting predicate elimination**
+   *
+   * .. math::
+   *   \inferrule{F, F_1 \dots F_n \mid G, (ids (ida (idr)?)?)?}{G}
+   *
+   * where :math:`\texttt{Rewriter}_{idr}(F \circ \sigma_{ids, ida}(F_n) \circ
```
[the remark this answers is not in the document]
Comment (nafur) [reply src/proof/proof_rule.h:217]: Yeah, I know. I didn't see an easy way to fix it...

File: src/proof/proof_rule.h [hunk 5]
```diff
@@ -37,606 +53,956 @@ namespace cvc5 {
...
-  // to apply, e.g. Rewriter::rewrite.
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Builtin theory -- Rewrite**
+   *
+   * .. math::
+   *   \inferrule{- \mid t, idr}{t = \texttt{Rewriter}_{idr}(t)}
+   *
+   * where :math:`idr` is a MethodId identifier, which determines the kind of
+   * rewriter to apply, e.g. Rewriter::rewrite. \endverbatim
```
[the remark this answers is not in the document]
Comment (nafur) [reply src/proof/proof_rule.h:129]: This is an interesting question. These methods generally point to internal functions that are not exposed (except indirectly via the proof rules). Maybe discuss this in the next proofs meeting?

File: src/proof/proof_rule.h [hunk 6]
```diff
@@ -37,606 +53,956 @@ namespace cvc5 {
...
-  // Arguments: none
-  // ---------------------
-  // Conclusion: (F2)
-  // Note this can optionally be seen as a macro for IMPLIES_ELIM+RESOLUTION.
+  /**
+   * \verbatim embed:rst:leading-asterisk
+   * **Boolean -- Modus Ponens**
+   *
+   * .. math::
+   *   \inferrule{F_1, (F_1 \implies F_2) \mid -}{F_2}
```
[the remark this answers is not in the document]
Comment (HanielB) [reply src/proof/proof_rule.h:563]: Yeah. The `\implies` is too big.

File: src/proof/proof_rule.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/proof/proof_rule.h:129]: Sounds good.

File: src/proof/proof_rule.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (HanielB) [reply src/proof/proof_rule.h:217]: Maybe leave it for now and later on we could use the same notation we used in the paper for rewriting. that would make this be shorter.

**Review by HanielB — APPROVED** [review]

**Review by ajreynol — APPROVED** [review]

**Pull request merged**